### PR TITLE
testnet-automation: More knobs for rolling upgrades

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

1.2 -> 1.3 still falls over

#### Summary of Changes

* Plumb genesis options to gain control over `--cluster-type`

### CODE REVIEW

**Comment by ryoqun** [comment]
@t-nelson Please take look at this: https://github.com/solana-labs/solana/pull/12390#issuecomment-696751381 :)

**Comment by t-nelson** [comment]
@ryoqun I've dropped the junk commits here and picked in the custom build commits

https://github.com/solana-labs/solana/pull/12337/commits/102f7d2dba4139180c805bcea8ec7101f20fb448 can be dropped once #12405 and #12406 land

https://github.com/solana-labs/solana/pull/12337/commits/fa525877dd3fcbca80fdb96e9fd3da0c4265cb06 may or may not be a good idea.  It seemed to work, but I'm not sure if it'll break MacOS/Windows

File: ci/forgive-me.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  rm -rf "$tmp_source_dir"
+}
+
+set -e
+tmp_source_dir="$(mktemp --directory --tmpdir="$source_dir")"
+
+trap 'cleanup_tmp_source_dir' EXIT
+
+# Clone the specified committish
+git clone "$remote" "$tmp_source_dir"
```
Comment (ryoqun) [inline ci/forgive-me.sh:22]: Is there any reason we have to clone all new? I think we can add another worktree like this:

```
# Create temporal source dir to build
git worktree add "$tmp_source_dir"
```

File: ci/forgive-me.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+here="$(realpath "$(dirname "$0")")"
+source_dir="${here:?}/.."
+
+committish="${1:?}"
+tarballBaseName="${2:-solana-release}"
+remote=https://github.com/solana-labs/solana.git
+
+cleanup_tmp_source_dir() {
+  rm -rf "$tmp_source_dir"
```
Comment (ryoqun) [inline ci/forgive-me.sh:13]: https://github.com/solana-labs/solana/pull/12337/files#r493804567 and do this?:

```
git worktree remove --force "$tmp_source_dir"
```

File: ci/forgive-me.sh (hunk 1 above)
Comment (t-nelson) [reply ci/forgive-me.sh:22]: Nope!  This was just a quick and dirty script to see if I could get what I wanted working (see script file name 😅).  Plenty of optimizations to be had here

File: ci/forgive-me.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+# Clone the specified committish
+git clone "$remote" "$tmp_source_dir"
+
+if pushd "$tmp_source_dir"; then
+  (
+    set +e
+
+    git fetch origin "$committish":branch_to_build
+    git checkout branch_to_build
```
Comment (ryoqun) [inline ci/forgive-me.sh:29]: this can be merged into https://github.com/solana-labs/solana/pull/12337/files#r493804567

File: ci/forgive-me.sh (hunk 1 above)
Comment (ryoqun) [reply ci/forgive-me.sh:22]: what a quick reply. :) I updated my comment by picking something from my attempt #12390. let's unify our work for the better. :)

Comment (ryoqun) [reply ci/forgive-me.sh:22]: this way we can build off at arbitrary commit, might be useful for building with fast-changing release branches.

File: scripts/cargo-install-all.sh [hunk 4]
```diff
@@ -109,12 +109,14 @@ done
 (
   set -x
   # shellcheck disable=SC2086 # Don't want to double quote $rust_version
-  $cargo $maybeRustVersion build $maybeReleaseFlag "${binArgs[@]}"
+  $cargo $maybeRustVersion build $maybeReleaseFlag --bins
```
Comment (ryoqun) [inline scripts/cargo-install-all.sh:112]: qq: why is this needed? because of shellcheck?

File: scripts/cargo-install-all.sh [hunk 5]
```diff
@@ -109,12 +109,14 @@ done
...
 )
 
 mkdir -p "$installDir/bin"
-for bin in "${BINS[@]}"; do
-  cp -fv "target/$buildVariant/$bin" "$installDir"/bin
+for bin in "target/$buildVariant/"*; do
```
Comment (ryoqun) [inline scripts/cargo-install-all.sh:116]: same as https://github.com/solana-labs/solana/pull/12337/files#r493815963

File: scripts/cargo-install-all.sh (hunk 4 above)
Comment (t-nelson) [reply scripts/cargo-install-all.sh:112]: IIRC some bins were missing without

File: scripts/cargo-install-all.sh (hunk 5 above)
Comment (t-nelson) [reply scripts/cargo-install-all.sh:116]: Idea here was basically to get rid of the manually specified bin list, so we don't have to maintain it
